airflow/bin/cli.py

```diff
@@ -361,13 +362,87 @@ def run(args, dag=None):
 
     task = dag.get_task(task_id=args.task_id)
     ti = TaskInstance(task, args.execution_date)
-    ti.refresh_from_db()
-    if ti.current_state() == "failed":
-        # This behavior is always correct (failed tasks should not run), but is implemented to specifically handle the
-        # case where a worker goes down, the celery task gets re-queued, and the scheduler kills the airflow task
-        # without knowing to revoke the task from the celery queue.
+    ti.refresh_from_db(include_job_id=True)
+    from airflow.utils.state import State
+    if ti.current_state() in [
+        State.FAILED,
+        State.SUCCESS,
+        State.UP_FOR_RETRY,
+        State.RUNNING,
+        State.SHUTDOWN,
+        State.SKIPPED
+    ]:
+        # Return if we're not in a valid state for running.
         return
 
+    # Here we're checking that the hostname is not None or the empty string.
+    if not args.raw:
+        # If the hostname of a task is set, another worker process previously attempted to perform this work item,
+        # and did not explicitly succeed or fail.
+        if ti.hostname:
+            log.info(
+                "For submitted task \"{}\" hostname is set to \"{}\" - checking to see if this task is a zombie yet.".format(
+                    ti.task_id, ti.hostname
+                )
+            )
+            # If a previous version of this task exists, we only want to run a new version if the old version zombies
+            # out.  If the previous version is not zombied out, it is still possible that the scheduler bounced, and
+            # incorrectly thought this task was an orphan, in which case the previous version will eventually reclaim
+            # it.
+            from airflow.jobs import LocalTaskJob as LJ
+            TI = TaskInstance
+            secs = conf.getint('scheduler', 'scheduler_zombie_task_threshold')
+            limit_dttm = datetime.utcnow() - timedelta(seconds=secs)
+            log.info("Task id is %s job id is %s", ti.task_id, ti.job_id)
+            latest_heartbeat = session.query(LJ).filter(LJ.id == ti.job_id).first().latest_heartbeat
+            if latest_heartbeat >= limit_dttm:
+
+                real_ti = session.query(TI).filter(
+                    TI.dag_id == ti.dag_id,
+                    TI.task_id == ti.task_id,
+                    TI.execution_date == ti.execution_date
+                ).with_for_update().first()
+
+                if real_ti.state == State.RUNNING:
```

Comment:
This is a razor-thin race condition since we return if we're in state running up above.  However, it is possible for the state to be set to running between our initial check, and our check where we acquire a lock.